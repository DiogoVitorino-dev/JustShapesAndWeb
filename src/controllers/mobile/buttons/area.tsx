import React from "react";
import { StyleSheet, ViewStyle } from "react-native";
import { Gesture, GestureDetector } from "react-native-gesture-handler";
import Animated, {
  useAnimatedStyle,
  useSharedValue,
  withTiming,
} from "react-native-reanimated";

import { ButtonProps } from "..";

import { View } from "@/components/shared";
import Colors from "@/constants/Colors";
import { Position } from "@/constants/commonTypes";

interface AreaProps extends ButtonProps {
  indicatorSize?: number;
  style?: ViewStyle;
}

export function AreaButton({ onPress, style, indicatorSize = 70 }: AreaProps) {
  const opacityIndicator = useSharedValue(0);
  const scaleIndicator = useSharedValue(1);
  const posIndicator = useSharedValue<Position>({ x: 0, y: 0 });

  const updateIndicatorPos = ({ x, y }: Position) => {
    "worklet";
    posIndicator.value = {
      x: x - indicatorSize / 2,
      y: y - indicatorSize / 2,
    };
  };

  const callOnPress = (pressing: boolean) => {
    "worklet";
    if (onPress) {
      onPress({ jumping: pressing });
    }
  };

  const tap = Gesture.Pan()
    .maxPointers(1)
    .shouldCancelWhenOutside(true)
    .onBegin(({ x, y }) => {
      updateIndicatorPos({ x, y });
      opacityIndicator.value = 1;
      scaleIndicator.value = withTiming(0.8);

      callOnPress(true);
    })
    .onUpdate(({ x, y }) => updateIndicatorPos({ x, y }))
    .onFinalize(() => {
      scaleIndicator.value = withTiming(1);
      opacityIndicator.value = 0;
      callOnPress(false);
    });

  const indicatorAnimatedStyle = useAnimatedStyle(() => ({
    opacity: opacityIndicator.value,
    top: posIndicator.value.y,
    left: posIndicator.value.x,
    transform: [
      {
        scale: scaleIndicator.value,
      },
    ],
  }));

  return (
    <GestureDetector gesture={tap}>
      <View style={[styles.area, style]}>
        <Animated.View
          style={[
            indicatorAnimatedStyle,
            styles.indicator,
            {
              width: indicatorSize,
              height: indicatorSize,
              borderRadius: indicatorSize / 2,
            },
          ]}
        />
      </View>
    </GestureDetector>
  );
}

const styles = StyleSheet.create({
  area: {
    width: "50%",
    height: "100%",
    position: "absolute",
  },
  indicator: {
    borderColor: Colors.control.button,
    borderWidth: 3,
    zIndex: 1000,
  },
});
