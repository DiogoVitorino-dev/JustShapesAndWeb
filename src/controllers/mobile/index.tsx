import React from "react";
import { StyleSheet, View, ViewStyle } from "react-native";
import { useAnimatedReaction, useSharedValue } from "react-native-reanimated";

import AreaButton from "./buttons/area";
import AnalogicJoystick from "./directional/analogic";
import { ControlData, ControlProp, Directional } from "../typesControllers";

interface MobileControlProp extends ControlProp {}

export interface JoystickData extends Directional {}

export interface JoystickProp {
  size?: number;
  velocity?: number;
  onMove?: (data: JoystickData) => void;
  containerStyle?: ViewStyle;
}

export default function MobileControl({ onMove, velocity }: MobileControlProp) {
  const jump = useSharedValue<boolean>(false);
  const data = useSharedValue<JoystickData>({ angle: 0, x: 0, y: 0 });

  const handleOnMove = (newData: JoystickData) => {
    "worklet";
    data.value = newData;
  };

  const handleOnJump = (pressed: boolean) => {
    "worklet";
    jump.value = pressed;
  };

  useAnimatedReaction(
    (): ControlData => ({ jumping: jump.value, ...data.value }),
    (current, previous) => {
      if (current !== previous) {
        onMove(current);
      }
    },
  );

  return (
    <View style={styles.container}>
      <AnalogicJoystick onMove={handleOnMove} velocity={velocity} />
      <AreaButton onPress={handleOnJump} />
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    width: "100%",
    flex: 1,
    flexDirection: "row",
    alignItems: "flex-end",
    justifyContent: "space-between",
  },
});
