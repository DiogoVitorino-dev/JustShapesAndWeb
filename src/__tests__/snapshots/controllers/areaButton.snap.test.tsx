import { render } from "@testing-library/react-native";

import { AreaButton } from "@/controllers/mobile/buttons";

describe("Area Button - Snapshot mobile control test", () => {
  it("should render correctly", () => {
    const tree = render(<AreaButton />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
