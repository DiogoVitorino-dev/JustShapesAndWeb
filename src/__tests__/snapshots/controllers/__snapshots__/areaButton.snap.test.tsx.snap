// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Area Button - Snapshot mobile control test should render correctly 1`] = `
<View
  collapsable={false}
  style={
    [
      undefined,
      [
        {
          "height": "100%",
          "position": "absolute",
          "width": "50%",
        },
        undefined,
      ],
    ]
  }
>
  <View
    animatedStyle={
      {
        "value": {
          "left": 0,
          "opacity": 0,
          "top": 0,
          "transform": [
            {
              "scale": 1,
            },
          ],
        },
      }
    }
    collapsable={false}
    style={
      {
        "borderColor": "rgb(20, 210, 255)",
        "borderRadius": 35,
        "borderWidth": 3,
        "height": 70,
        "left": 0,
        "opacity": 0,
        "top": 0,
        "transform": [
          {
            "scale": 1,
          },
        ],
        "width": 70,
        "zIndex": 1000,
      }
    }
  />
</View>
`;
